src/index.py

```diff
@@ -3,16 +3,33 @@
 
 """Execute the uploading of documentation."""
 
+import itertools
+import re
+import typing
 from pathlib import Path
 
+from more_itertools import peekable
+
 from .constants import (
+    DOC_FILE_EXTENSION,
     DOCUMENTATION_FOLDER_NAME,
     DOCUMENTATION_INDEX_FILENAME,
     NAVIGATION_TABLE_START,
 )
 from .discourse import Discourse
-from .exceptions import DiscourseError, ServerError
-from .types_ import Index, IndexFile, Metadata, Page
+from .exceptions import DiscourseError, InputError, ServerError
+from .types_ import Index, IndexContentsListItem, IndexFile, Metadata, Page
+
+CONTENTS_HEADER = "# contents"
+CONTENTS_END_LINE_PREFIX = "#"
+
+_WHITESPACE = "( *)"
+_LEADER = r"((\d\.)|(\*)|(-))"
```

Comment:
This also only supports numbering up to 9 not more

Reply:
It also supports mixing leaders like:
```
* [title 1](value 1)
- [title 2](value 2)
1. [title 3](value 3)
```

We should probably enforce coherence over the leader choice on the hierarchy

Reply:
the first two points are good ones, coherence is up to the user. For example, 

```markdown
1. ...
  - ...
```

Is reasonable. Enforcing coherence would be quite complicated and probably better suited to a markdown linter

Reply:
first two point should be resolved in the next push

Reply:
Asking the user to ensure coherence is fair, but I feel like it would be better to either lint indeed the contents index, or add some more checks (ensuring the same leader in the same hierarchy level shouldn't be too complicated, but ensuring letter and numbers are following the previous one probably more complex, but the recursive approach will help implementing this check)
Perhaps for a further task ?